src/signatures.rs

```diff
@@ -202,57 +344,94 @@ fn read_expected(bytes: &mut &[u8], expected: u32) -> ParseResult<bool> {
     })
 }
 
+fn read_string<'a>(bytes: &mut &'a [u8]) -> &'a str {
+    let length = read_unsigned(bytes).unwrap();
+    let (string_bytes, rest) = bytes.split_at(length as usize);
+    *bytes = rest;
+    std::str::from_utf8(string_bytes).unwrap()
+}
+
+fn read_i8(bytes: &mut &[u8]) -> i8 {
+    let (value_bytes, rest) = bytes.split_at(std::mem::size_of::<i8>());
+    *bytes = rest;
+    i8::from_le_bytes(value_bytes.try_into().unwrap())
+}
+
+fn read_u8(bytes: &mut &[u8]) -> u8 {
```

Comment:
Not too happy with these - seems very repetitive, but couldn't figure out how to write this as a generic function - much like a function template in C++.

Reply:
You could use a macro to write these.